Fix: undefined index warning.

ankitrox opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #7801 

## Relevant technical choices

<!-- Please describe your changes. -->

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 5.2 and PHP 7.4.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.
- [x] Ensure there are no unexpected significant changes to file sizes.

## Merge Reviewer Checklist

- [x] Ensure the PR has the correct target branch.
- [x] Double-check that the PR is okay to be merged.
- [x] Ensure the corresponding issue has a ZenHub release assigned.
- [x] Add a changelog message to the issue.

techanvil commented on tests/phpunit/integration/Modules/Analytics_4Test.php:
```diff
@@ -2923,6 +2923,52 @@ public function test_create_custom_dimension() {
+	public function test_create_custom_dimension_without_optional_params() {
+		$property_id = '123456789';
+
+		$this->fake_handler_and_invoke_register_method( $property_id );
+
+		$this->grant_scope( Analytics_4::EDIT_SCOPE );
+
+		$custom_dimension = array(
+			'displayName'   => 'Test Custom Dimension',
+			'parameterName' => 'googlesitekit_post_author',
+			'scope'         => 'EVENT',
+		);
+
+		$response = $this->analytics->set_data(
+			'create-custom-dimension',
+			array(
+				'propertyID'      => $property_id,
+				'customDimension' => $custom_dimension,
+			)
+		);
+
+		$this->assertNotWPError( $response );
+
+		$response_array = (array) $response;
+
+		// Assert that the keys exist.
+		$keys = array_keys( $custom_dimension );
+
+		foreach ( $keys as $key ) {
+			$this->assertArrayHasKey( $key, $response_array );
+		}
+
+		// Validate the response against the expected mock value.
+		foreach ( $custom_dimension as $key => $value ) {
+			$this->assertEquals( $value, $response_array[ $key ] );
+		}
```
Asserting the response like this doesn't quite make sense, because it's hardwired to be an object which _does_ have the optional fields set:

https://github.com/google/site-kit-wp/blob/66c60822f669d225af698f1fb4aad769695a2e63/tests/phpunit/integration/Modules/Analytics_4Test.php#L3139-L3151

I'd suggest reorganising the test as follows:

```suggestion
		// Create a custom dimension with only the required fields present.
		$raw_custom_dimension = array(
			'displayName'   => 'Test Custom Dimension',
			'parameterName' => 'googlesitekit_post_author',
			'scope'         => 'EVENT',
		);

		$this->fake_handler_and_invoke_register_method(
			$property_id,
			function ( Request $request ) use ( $raw_custom_dimension, $property_id ) {
				$url = parse_url( $request->getUri() );

				if ( "/v1beta/properties/$property_id/customDimensions" === $url['path'] ) {
					$custom_dimension = new GoogleAnalyticsAdminV1betaCustomDimension( $raw_custom_dimension );

					return new Response(
						200,
						array(),
						json_encode( $custom_dimension )
					);
				}

				return new Response( 200 );
			}
		);

		$this->grant_scope( Analytics_4::EDIT_SCOPE );

		$response = $this->analytics->set_data(
			'create-custom-dimension',
			array(
				'propertyID'      => $property_id,
				'customDimension' => $raw_custom_dimension,
			)
		);

		$this->assertNotWPError( $response );

		$this->assertEquals( $raw_custom_dimension, (array) $response->toSimpleObject() );
```

See the `test_sync_audiences()` function in the same file for reference.

techanvil commented on tests/phpunit/integration/Modules/Analytics_4Test.php:
```diff
@@ -2923,6 +2923,52 @@ public function test_create_custom_dimension() {
 		$this->assertEquals( "/v1beta/properties/$property_id/customDimensions", $request_url['path'] );
 	}
 
+	public function test_create_custom_dimension_without_optional_params() {
```
This test name is a bit confusing when compared with the existing `test_create_custom_dimension__required_params()`:

https://github.com/google/site-kit-wp/blob/66c60822f669d225af698f1fb4aad769695a2e63/tests/phpunit/integration/Modules/Analytics_4Test.php#L2781

`test_create_custom_dimension__required_params()` is testing the required parameters for the whole request whereas this test is validating the requirements for the custom dimension.

I'd suggest renaming along these lines:

```suggestion
	public function test_create_custom_dimension__without_optional_fields() {
```

techanvil reviewed (changes requested):
Thanks @ankitrox, this is working well in my testing.

I've left a suggestion about the test case.

Also, the QAB is a bit unclear - for one thing, it looks like the error will only surface with PHP 8+ so this should be made clear. Secondly the QAB is describing how to reproduce the error, whereas the expectation is that the error will no longer occur. Please rephrase it for clarity.

ankitrox commented:
Thank you @techanvil , it was my bad about the QAB that I added during testing and added instructions such that it should test for errors rather than testing that errors should not be present. 🤦🏼 I have corrected it now and also made it clear to setup the site with php 8+. Thanks for pointing this out.

Also, made the changes in test as per couple of suggestions provided by you.

Thank you for reviewing the PR. Assigning this to you again to review the changes.

techanvil reviewed (commented):
Thanks @ankitrox! This LGTM.

Please note I pushed a minor tweak to add some newlines to aid readability.

:white_check_mark:
